### ISSUE TITLE
Implement dependency model

### ISSUE THREAD

**Issue by jvalkeal** [spring-cloud/spring-cloud-dashboard#10]
It would be very handy to have dependency model for applications. Stream already have a naive deps as sink is started first, then processors and lastly source. One huge issue with this is that how we store vertex edge information which is very cumbersome with relational DB's. Doable but rather complicated.

I tried https://github.com/jgrapht/jgrapht which seem to be rather nice to create graph models and then traverse it with different iterators.

Not sure if `neo4j` is ever going to be an option.

**Comment by markfisher** [comment]
Before we go too far down this path, I think it's worth considering the requirement in light of this article by @mstine: https://medium.com/built-to-adapt/on-orchestrated-microservices-5c8bd787ead4

**Comment by jvalkeal** [comment]
Yeah he certainly have a good point :) My head was already spinning when having a thought how relations could be stored in a database https://inviqa.com/blog/graphs-database-sql-meets-social-network.

**Comment by mbogoevici** [comment]
The other questionable aspect is that it is hard to guarantee that deployments actually happen in the specified order - i.e. it is safer to assume that the actual deployments (e.g. resource allocation, startup, etc, happen asynchronously).

**Comment by jvalkeal** [comment]
Well, main focus of this dep model was not to guarantee that apps get into running state on a specific order as realistically we could only try to start on a specific order and possibly stop further deployments if app deploy fails.

Point was to get relational model for all kind of apps. I.e. if we deploy stream we'd have deps(we already hardcode this starting order) `sink` -> `processor` -> `source` and starting one of those would then automatically start all. Over the top if you have just one stream but having a complex combination of apps/streams one could create a DAG of 10's of streams and apps. Then this dep model would orchestrate start of this whole group(or whatever we want to call it).

Kinda played this with `jgrapht`
```
DefaultDirectedGraph<String, DefaultEdge> main =
		DefaultDirectedGraph.<String, DefaultEdge>builder(DefaultEdge.class)
			.addVertices("source11", "processor11", "sink11")
			.addVertices("source21", "sink21")
			.addVertices("source31", "sink31")
			.addVertices("app11", "app21")
			.addVertices("app31", "app41", "app51")
			.addEdgeChain("source11", "processor11", "sink11")
			.addEdgeChain("source21", "sink21")
			.addEdgeChain("source31", "sink31")
			.addEdge("app21", "app11")
			.addEdge("app21", "source21")
			.addEdge("source21", "source11")
			.addEdge("source21", "source31")
			.addEdge("app41", "app51")
			.build();

Iterator<String> dfs = new TopologicalOrderIterator<>(main);
	while (dfs.hasNext()) {
	System.out.println(dfs.next());
}
```
which prints
```
app21
app31
app41
app11
source21
app51
sink21
source11
source31
processor11
sink31
sink11
```
model can even be disconnected graph as in above `app31` is a standalone vertex and `app41` only connects to `app51`.
